File: woodroid/src/main/java/me/gilo/woodroid/data/api/ProductAPI.java
package me.gilo.woodroid.data.api;


import me.gilo.woodroid.models.Product;
import retrofit2.Call;
import retrofit2.http.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public interface ProductAPI {


    @GET("products")
    Call<ArrayList<Product>> getProducts();

    @GET("products/{id}")
    Call<Product> getProduct(@Path("id") int id);

    @GET("products")
    Call<ArrayList<Product>> getProducts(@Query("filter[category]") String category);

    @GET("products")
    Call<ArrayList<Product>> search(@Query("search") String search);

    @GET("products")
    Call<List<Product>> filter(@QueryMap Map<String, String> filter);

    @GET("products/count")
    Call<List<Product>> getProductsCount();


    @Headers("Content-Type: application/json")
    @POST("products")
    Call<Product> create(@Body Product body);

    @GET("products/{id}")
    Call<Product> view(@Path("id") int id);

    @GET("products")
    Call<List<Product>> list();

    @Headers("Content-Type: application/json")
    @PUT("products/{id}")
    Call<Product> update(@Path("id") int id, @Body Product body);

    @DELETE("products/{id}")
    Call<Product> delete(@Path("id") int id);

    @DELETE("products/{id}")
    Call<Product> delete(@Path("id") int id, @Query("force") boolean force);

    @POST("products/batch")
    Call<String> batch(@Body Product body);

}

File: woodroid/src/main/java/me/gilo/woodroid/data/api/ProductAPI.kt
package me.gilo.woodroid.data.api


import me.gilo.woodroid.models.Product
import retrofit2.Call
import retrofit2.http.*

import java.util.ArrayList

interface ProductAPI {


    @get:GET("products")
    val products: Call<ArrayList<Product>>

    @get:GET("products/count")
    val productsCount: Call<List<Product>>

    @GET("products/{id}")
    fun getProduct(@Path("id") id: Int): Call<Product>

    @GET("products")
    fun getProducts(@Query("filter[category]") category: String): Call<ArrayList<Product>>

    @GET("products")
    fun search(@Query("search") search: String): Call<ArrayList<Product>>

    @GET("products")
    fun filter(@QueryMap filter: Map<String, String>): Call<List<Product>>


    @Headers("Content-Type: application/json")
    @POST("products")
    fun create(@Body body: Product): Call<Product>

    @GET("products/{id}")
    fun view(@Path("id") id: Int): Call<Product>

    @GET("products")
    fun list(): Call<List<Product>>

    @Headers("Content-Type: application/json")
    @PUT("products/{id}")
    fun update(@Path("id") id: Int, @Body body: Product): Call<Product>

    @DELETE("products/{id}")
    fun delete(@Path("id") id: Int): Call<Product>

    @DELETE("products/{id}")
    fun delete(@Path("id") id: Int, @Query("force") force: Boolean): Call<Product>

    @POST("products/batch")
    fun batch(@Body body: Product): Call<String>

}